react_juce/core/TextInputView.cpp

```diff
@@ -0,0 +1,207 @@
+/*
+==============================================================================
+
+  TextInputView.cpp
+  Created: 20 Jan 2021 10:30pm
+
+==============================================================================
+*/
+
+#include "TextInputView.h"
+#include "TextView.h"
+
+namespace blueprint
+{
+    namespace detail
+    {
+        static juce::var makeInputEventObject(const juce::String &value)
+        {
+            juce::DynamicObject::Ptr obj = new juce::DynamicObject();
+            obj->setProperty("value", value);
+            return obj.get();
+        }
+
+        static juce::var makeChangeEventObject(const juce::String &value)
+        {
+            juce::DynamicObject::Ptr obj = new juce::DynamicObject();
+            obj->setProperty("value", value);
+            return obj.get();
+        }
+    }
+
+    void TextInput::setControlledValue(const juce::String &value)
+    {
+        insertedAsControlledValue = true;
+        setText(value);
+    }
+
+    void TextInput::setMaxLength(int maxLen)
+    {
+        maxLength = maxLen;
+        setInputRestrictions(maxLen);
+    }
+
+    void TextInput::setPlaceholderText(const juce::String &text)
+    {
+        placeholderText = text;
+        setTextToShowWhenEmpty(placeholderText, placeholderColour);
+    }
+
+    void TextInput::setPlaceholderColour(const juce::Colour &colourToUse)
+    {
+        placeholderColour = colourToUse;
+        setTextToShowWhenEmpty(placeholderText, placeholderColour);
+    }
+
+    void TextInput::insertTextAtCaret(const juce::String &textToInsert)
+    {
+        const juce::String currentValue = getText();
+        juce::TextEditor::insertTextAtCaret(textToInsert);
+        const juce::String newValue = getText();
+        if (currentValue == newValue)
+        {
+            return;
+        }
+
+        // Invoke JavaScript's `input` event.
+        if (props.contains(TextInputView::onInputProp) && props[TextInputView::onInputProp].isMethod())
+        {
+            std::array<juce::var, 1> args{{detail::makeInputEventObject(newValue)}};
+            juce::var::NativeFunctionArgs nfArgs(juce::var(), args.data(), static_cast<int>(args.size()));
+            std::invoke(props[TextInputView::onInputProp].getNativeFunction(), nfArgs);
+        }
+
+        dirty = true;
+
+        if (controlled && !insertedAsControlledValue)
```

Comment:
Wondering if we can move this check to the top of the function? Seems that we might be able to just return early and no-op if we know that we're a controlled component? It seems odd that we will have just invoked `onInput` event too if we know we're going to undo it immediately after, but maybe that matches ReactDOM's behavior anyway?

Reply:
Thanks @nick-thompson for detailed review!
And I'm OK for making it better!

I want make sure what you're saying.
Do you want me to change the code like this?

```cpp
    void TextInput::insertTextAtCaret(const juce::String &textToInsert)
    {
        if (controlled && !insertedAsControlledValue)
            return;

        // same as before

        undo();
        insertedAsControlledValue = false;
    }
```

If so, I don't agree with the idea.
If we adopt it, input event will never be invoked for controlled components and there will be no way to know characters entered in the controlled text editor.

I might misunderstand what you say.
If so, please let me know!

Reply:
Ohh of course, of course. I misunderstood, you're right!

Let's keep it as is, I'll go ahead and merge :)

Thank you again!